---
title: "Gene Set Pathway and Network Analysis by Using GSEA, Cytoscape and Enrichment Map"
author: "Xinyi Zhang"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_depth: 3
bibliography: A3_XinyiZhang.bib
---

## Loading Packages and Data From A2
```{r library_loading, message=FALSE, warning=FALSE, results='hide'}

if (!require("knitr", quietly = TRUE))
  install.packages("knitr")
if (!require("RSQLite", quietly = TRUE)){
  install.packages("RSQLite")
}

library(knitr)
library(RSQLite)
load("A3.RData")
```

The ranked gene list from A2 is shown below. There are multiple parameters that can be selected. We selected to use the up-regulated rank by logFC, to keep same with the original paper. Original data GSE_150392 can be downloaded from https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150392

```{r}
knitr::kable(MultipuleRank, caption = 'Table 1. Ranked Gene list by logFC Up-Regulated')
```

## Non-thresholded Gene Set Enrichment Analysis

Conduct non-thresholded gene set enrichment analysis using the ranked set of genes from Assignment #2.

#### 1.What method did you use? What genesets did you use? Make sure to specify versions and cite your methods.

I used GSEA4.2.3 in this part; Cytoscape 3.9.1, its Enrichment Map application and AutoAnnotation (1.3.5) application for the downstream analysis.

#### 2.Summarize your enrichment results.

```{r}
knitr::kable(ranked_gene_list_na_pos_versus_na_neg_1648855165869, caption = 'Table 2. Ranked Gene List Virus vs. Mock')
```

* 16035 gene sets with size larger than 200 or smaller than 15 were filtered out from a total of 16746, the remaining 711 gene sets were used in this analysis. We used a stricter boundary to obtain more rigorous results. We selected the permutation number of 1000, which has a suitable running time for the program.

* The Positive Phenotype (na_pos):

![Table 3. Enrichment Plot of the Top 20 Gene Sets Ranked in the Positive(Virus) Phenotype](./A3/GSEA/pos.JPG)

517 out of 711 gene sets are returned in the Virus (na_pos) sub-type. 
421 gene sets are significant at FDR < 25%.
304 gene sets are significantly enriched at nominal p-value < 1%.
366 gene sets are significantly enriched at nominal p-value < 5%.
The most related gene set in this sub-type according to the analysis is DEFENSE RESPONSE GOBP GO:0006952, with an enrichment score (ES) of 0.5994537 and a normalized enrichment score (NES) of 4.323445.

![Figure 1. Profile of the Running ES Score & Positions of DEFENSE RESPONSE GOBP GO:0006952 on the Rank Ordered List](./A3/GSEA/Virus_Top.png)

![Figure 2. Random ES Distribution of DEFENSE RESPONSE GOBP GO:0006952](./A3/GSEA/Virus_Top_RandomESDistribution.png)

```{r}
knitr::kable(DEFENSE_RESPONSE_GOBP, caption = 'Table 4. DEFENSE RESPONSE GOBP')
```

* The Negative Phenotype (na_neg):

![Table 5. Enrichment Plot of the Top 20 Gene Sets Ranked in the negative(Mock) Phenotype](./A3/GSEA/neg.JPG)

194 out of 711 gene sets are returned in the Mock (na_neg) sub-type. 
98 gene sets of them are significantly enriched at FDR < 25%. 
69 gene sets of them are significantly enriched at nominal p-value < 1%. 
89 gene sets of them are significantly enriched at nominal p-value < 5%.
The most related gene set in this sub-type according to the analysis is CELL CYCLE REACTOME R-HSA-1640170.1, with an enrichment score (ES) of -0.56024605 and a normalized enrichment score (NES) of -3.6521144.

![Figure 3. Profile of the Running ES Score & Positions of CELL CYCLE REACTOME R-HSA-1640170.1 on the Rank Ordered List](./A3/GSEA/Mock_Top.png)

![Figure 4. Random ES Distribution of CELL CYCLE REACTOME R-HSA-1640170.1](./A3/GSEA/Mock_Top_RandomESDistribution.png)

```{r}
knitr::kable(CELL_CYCLE_REACTOME, caption = 'Table 6. CELL CYCLE REACTOME')
```

* However, this sample contains only 831 objects, which might be too few for the GSEA analysis.


#### 3.How do these results compare to the results from the thresholded analysis in Assignment #2. Compare qualitatively. Is this a straight forward comparison? Why or why not?

These results do not have a threshold which is a cut-off according to p-value, so the returned gene sets are slightly different. It is a straightforward comparison. Because they are both imported from the same normalized data.

## Visualize your Gene set Enrichment Analysis in Cytoscape

Visualizing the results from the non-thresholded gene set enrichment analysis in Cytoscape by Using Enrichment Map and AutoAnnotate.

#### 1. Create an enrichment map - how many nodes and how many edges in the resulting map? What thresholds were used to create this map? Make sure to record all thresholds. Include a screenshot of your network prior to manual layout.

* In the Virus vs Mock gene set enrichment map visualization, we used FDR q-value of 0.01, p-value of 0.1 and edge similarity value of 0.375 as cutoff. We selected all NES. The Metric was set to 50% Jaccard combined with 50% Overlap. There are 339 nodes and 3033 edges in this visualized enrichment map.

![Figure 5. Pre-Analysis Parameters](./A3/Cytoscape/pre-analysis.png)

* Below is the over view of the network.

![Figure 6. Overview of the Network](./A3/Cytoscape/overview.png)

#### 2. Annotate your network - what parameters did you use to annotate the network. If you are using the default parameters make sure to list them as well.

* We used the default settings of the AutoAnnotate application to annotate the entire map, we selected to layout network to prevent the overlap of the clusters. We used GS_DESCR as label column. 

![Figure 7. Pre-Annotation Parameters](./A3/Cytoscape/pre-annotation parameters.png)

* We selected a round shape, border width of 3, opacity of 20%, font scale of 20% and minimum font size of 4.

![Figure 8. Annotation Parameters](./A3/Cytoscape/annotation parameters.png)

#### 3. Make a publication ready figure - include this figure with proper legends in your notebook.

![Figure 9. Mock vs. Virus Enrichment Map](./A3/Cytoscape/publication_ready_figure.png)


#### 4. Collapse your network to a theme network. What are the major themes present in this analysis? Do they fit with the model? Are there any novel pathways or themes?

* The major themes that fit with the model include blood circulation heart,actin contraction filament, organism organelle organization, mitotic cycle sister, signaling cellular interferon, apoptotic programmed death, naba factor matrisome, positive cell proliferation, response biotic virus, taxis migration chemotaxis, programmed cell death, component movement migration, hallmark allograft rejection, activation involved neutrophil, viral life multi, negative regulation cellular, regulation metabolic positive, inflammation response stress, vasculature development, protein localization transport, transcription regulation biosynthetic, transmembrane transporter activity.

* There are also novel pathways and themes that do not fit with the model: epithelial mesenchymal, myogenesis, tnfa signaling, inflammatory response, hypoxia, muscel development tissue, extracellular organization matrix, estrogen response, oxidative phosphorylation, e2f targets, p53 pathway, interferon gamma alpha, uv response, kras signaling, g2m checkpoint and il2 stat5.

![Figure 10. Mock vs. Virus Collapsed Theme Network](./A3/Cytoscape/collapsed publication ready figure.png)

## Interpretation and detailed view of results

#### 1. Do the enrichment results support conclusions or mechanism discussed in the original paper? How do these results differ from the results you got from Assignment #2 thresholded methods

Yes, the enrichment results support conclusions and mechanisms discussed in the original paper at some point. 

The original paper concluded that SARS-CoV-2 infection (Virus group) induces significant transcriptional changes within hiPSC-CMs(human iPSC cardiomyocytes in vitro) in gene pathways related to cellular metabolism and immune response. 

In the Virus group, considerable amount of gene sets and themes are related to reacting to abnormality according to the positive phenotype enrichment plot (Table 3.) and the enrichment map (Figure 9.), such as virus response, inflammatory response, defense response, mass of apoptosis, etc. There are also many themes related to metabolic, transcriptional and cellular pathways. 

But we didn't see any significant themes about normal heart activities as those shown in the Mock group. So we conclude that the SARS-CoV-2 infection might lead to immune response and affect normal heart activity in metabolic, transcriptional and cellular pathways, such as functional alterations, transcriptional changes, apoptosis mentioned in the result of original paper.


#### 2. Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your result?

* Yes. Compared to the in vitro experiment conducted by Sharma et al., Kumar et al. performed a study on well-characterized cardiomyocytes from lymphoblastoid cell line (LCL) reprogrammed iPSC lines of 12 participants belong to four sibships. The samples were analyzed by genome-wide RNA sequencing based transcriptomic and functional annotation. (Kumar et al., 2021)

In the enrichment analysis, the CM-upregulated transcriptome showed significantly high enrichment in the gene set GSE150392 (the original data for this assignment) down-regulated by SARS-CoV-2 infection in cardiomyocytes. (Kumar et al., 2021)

![Figure S1. (Kumar et al., 2021)](./A3/Inteperation/Kumar.png)

* Perez-Bermejo et al. also indicated that SARS-CoV-2 can productively infect human cardiomyocytes inducing broad transcriptional changes, disrupt intracellular features after using iPSC-derived tissues to analyze to evaluate the SARS-CoV-2 infectivity of different heart cell types by using scRNA-sequencing and immunofluorescence.

They observed that SARS-CoV-2 infection creates precise and ordered disruptions to the myofibrillar structure and dissolution of the cardiac contractile machinery, which would lead to cardiac functional collapse. (Perez-Bermejo et al., 2021)

#### Add a post analysis to your main network using specific transcription factors, microRNAs or drugs. Include the reason why you chose the specific miRs, TFs or drugs (i.e publications indicating that they might be related to your model). What does this post analysis show?

I will choose Troponin I to study, since the increase of Troponin I is a clinical biomarkers of cardiac damage as well as the strongest predictors of mortality in COVID-19 patients. (Perez-Bermejo et al., 2021)

Troponin I (TnI) is the protein subunit that inhibits muscle contraction in the absence of calcium. (Roos et al., 1997) The interaction of troponin I with actin is weakened in the presence of Ca2+. Troponin I then move from actin to a larger distance and the steric blocking of the actin-myosin interaction vanishes. (Filatov et al., 1999)

It is not surprising that the increase of Troponin I lead to cardiac damage, since it is the inhibitor of muscle contraction. It is also possible that the level of Ca2+ could have further impact on the SARS-CoV-2 influenced cardiomyocytes infection and cardiac functional damage, since the inhibition of troponin I on muscle contraction increases when the Ca2+ level decreases. This can be studied in the future.

## Reference

(A3_XinyiZhang.bib)
